actions/release-build-publish/action.yml

```diff
@@ -7,15 +7,50 @@ inputs:
   gpg_passphrase:
     description: 'GPG passphrase for artifact signing'
     required: true
+  update_type:
+    description: 'One of MAJOR, MINOR, BUILD, PATCH'
+    required: true
 
 runs:
   using: "composite"
   steps:
-    - name: Get version
-      id: get_version
+    - name: Get old version
+      id: get_old_version
       shell: bash
       run: |
         echo "version=$(python build/versionutils.py gradle.properties)" >> "$GITHUB_OUTPUT"
+    # Push a version bump back to main. There are failure scenarios that can result
+    # in published artifacts but an erroneous build, so it's safer to bump the version
+    # at the beginning
+    - name: Increment version
+      run: python build/versionutils.py gradle.properties --increment --commit -u ${{ inputs.update_type }}
+    - name: Get new version
+      id: get_new_version
+      shell: bash
+      run: |
+        echo "version=$(python build/versionutils.py gradle.properties)" >> "$GITHUB_OUTPUT"
+
+    # Create and push the tag
+    - name: Create tag
+      shell: bash
+      run: git tag -m "Release ${{ steps.get_new_version.outputs.version }}" -f "${{ steps.get_new_version.outputs.version }}"
```

Comment:
I think we should move this to after the updating the release notes. If the release notes script fails, then nothing gets pushed (artifacts or commits), so no harm, no foul. However, at the end of the file (post #3078), we check out the tag at the very end, which we need so that we're running from the correct place when we build the documentation. At present, that would mean that the release notes file (published as part of the documentation step) would be missing the most recent update when we go to the docs, and it won't get published until the subsequent release.

Reply:
Maybe that means that the script needs to have a mode where instead of comparing tags to get the list of commits, it needs to compare the old tag to the current HEAD? As I think the script now will fail as the tag hasn't actually been pushed anyway